### PR TITLE
[RPA-821] Limit creating and executing processes by guests

### PR DESCRIPTION
**Pull request by star-in-the-jar**
## Description

- translacje
- liczenie UDANYCH uruchomień procesu
- blokowanie tworzenia nowego procesu w przypadku posiadania już jednego przez guest'a
image

image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

---

I agree to the terms of the RunBotics Contributor License Agreement.

### CODE REVIEW

**Review by star-in-the-jar — COMMENTED** [review]
fast self review

**Review by star-in-the-jar — COMMENTED** [review]
self reviewed

File: runbotics-orchestrator/src/main/java/com/runbotics/service/impl/ProcessServiceImpl.java [hunk 1]
```diff
@@ -215,6 +216,17 @@ public Optional<ProcessDTO> findOne(Long id) {
...
+        log.debug("Request to check if current user can create process");
+        var currentUser = this.userService.getUserWithAuthorities().get();
+        var isGuest = currentUser.getAuthorities().contains(createGuestAuthority());
+        List<Process> userProcesses = processRepository.findByCreatedByUser(currentUser.getLogin(), currentUser.getLogin());
+        return !isGuest || userProcesses.size() == 0;
+    }
+
+
+
+
```
Comment (preludium) [inline runbotics-orchestrator/src/main/java/com/runbotics/service/impl/ProcessServiceImpl.java]: image

File: runbotics-orchestrator/src/main/java/com/runbotics/service/ProcessService.java [hunk 2]
```diff
@@ -57,6 +57,13 @@ public interface ProcessService {
      */
     Optional<ProcessDTO> findOne(Long id);
 
+    /**
+     * Check if user is a guest without any processes or simply is not a guest.
+     *
+     * @return false if is a guest with any process.
+     */
+    boolean getCanBeCreatedByCurrentUser();
```
Comment (preludium) [inline runbotics-orchestrator/src/main/java/com/runbotics/service/ProcessService.java]: imo `get` prefix is suited for getters in entity classes for example. Here you return boolean so it opens up a lot of possibilities to name it. I suggest `hasRequesterCreateProcessAccess`

File: runbotics-orchestrator/src/main/java/com/runbotics/web/rest/ProcessResource.java [hunk 3]
```diff
@@ -62,10 +66,17 @@ public ProcessResource(ProcessService processService, ProcessRepository processR
     @PostMapping("/processes")
     public ResponseEntity<ProcessDTO> createProcess(@Valid @RequestBody ProcessDTO processDTO) throws URISyntaxException {
         log.debug("REST request to save Process : {}", processDTO);
+        boolean canBeCreated = processService.getCanBeCreatedByCurrentUser();
+
+        if(!canBeCreated) {
+            throw new BadRequestAlertException("Guest can create only one process", ENTITY_NAME, "guestprocesslimit");
```
Comment (preludium) [inline runbotics-orchestrator/src/main/java/com/runbotics/web/rest/ProcessResource.java]: imo here should be access denied 403 type of error

File: runbotics/runbotics-common/src/model/api/guest.model.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+export interface Guest {
+    ip: string;
+    userId: number;
+    executionCount: number;
```
Comment (preludium) [inline runbotics/runbotics-common/src/model/api/guest.model.ts]: execution**s**Count*

File: runbotics-orchestrator/src/main/java/com/runbotics/web/rest/ProcessResource.java (hunk 3 above)
Comment (preludium) [reply runbotics-orchestrator/src/main/java/com/runbotics/web/rest/ProcessResource.java]: I created a new endpoint in the meanwhile to create processes for guests specifically so handle it also

File: runbotics/runbotics-orchestrator-ui/src/src-app/translations/en/components.json [hunk 5]
```diff
@@ -6,6 +6,7 @@
 
     "Component.BotProcessRunner.Warning": "Looking for a bot",
     "Component.BotProcessRunner.Error": "All bots are busy",
+    "Component.BotProcessRunner.Error.GuestRunLimitExceeded": "Guests have limited runs. Register to get unlimited executions.",
```
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/translations/en/components.json]: maybe better `You have reached your daily limit. Create an account to continue or come back tomorrow`

File: runbotics/runbotics-scheduler/src/database/guest/guest.entity.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+@Entity({ name: 'guest' })
+export class GuestEntity implements Guest {
+    @PrimaryColumn({ name: 'ip', type: 'varchar', length: 100 })
+        ip: string;
+
+    @Column({ name: 'user_id', type: 'bigint'})
+        userId: number;
+
+    @Column({ name: 'executions_count', type: 'int' })
+        executionCount: number;
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/database/guest/guest.entity.ts]: adjust field name as mentioned earlier

File: runbotics/runbotics-scheduler/src/database/guest/guest.service.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        return this.guestRepository.update({ userId }, { executionCount: newCount });
+    }
+
+    async incrementExecutionsCount(userId: number): Promise<UpdateResult> {
+        return this.guestRepository.increment({ userId }, 'executionCount', 1);
+    }
+
+    async decrementExecutionsCount(userId: number): Promise<UpdateResult> {
+        return this.guestRepository.decrement({ userId }, 'executionCount', 1);
+    }
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/database/guest/guest.service.ts]: if expect the user entity to be injected to the guest objects here then you need to include a relation in each repository method. See other services for examples

File: runbotics/runbotics-scheduler/src/database/user/user.module.ts [hunk 8]
```diff
@@ -1,11 +1,11 @@
...
+import { TypeOrmModule } from '@nestjs/typeorm';
+import { UserEntity } from './user.entity';
+import { UserService } from './user.service';
+
+@Module({
+    imports: [TypeOrmModule.forFeature([UserEntity])],
+    exports: [UserService],
+    providers: [UserService]
+})
+export class UserModule {}
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/database/user/user.module.ts:11]: lf to crlf?

File: runbotics/runbotics-scheduler/src/database/guest/guest.service.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+@Injectable()
+export class GuestService {
+    constructor(
+        @InjectRepository(GuestEntity)
+        private guestRepository: Repository<GuestEntity>,
+    ) { }
+
+    async findById(userId: number): Promise<Guest> {
+        return this.guestRepository.findOne({ where: { userId }});
+    }
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/database/guest/guest.service.ts]: include in the name that you mean userId

**Review by preludium — CHANGES_REQUESTED** [review]

File: runbotics-orchestrator/src/main/java/com/runbotics/service/impl/ProcessServiceImpl.java (hunk 1 above)
Comment (star-in-the-jar) [reply runbotics-orchestrator/src/main/java/com/runbotics/service/impl/ProcessServiceImpl.java]: xDDDDD

File: runbotics/runbotics-scheduler/src/database/guest/guest.service.ts (hunk 9 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/database/guest/guest.service.ts]: done

File: runbotics/runbotics-scheduler/src/database/user/user.module.ts (hunk 8 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/database/user/user.module.ts:11]: more like crlf to lf

File: runbotics/runbotics-orchestrator-ui/src/src-app/translations/en/components.json (hunk 5 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-orchestrator-ui/src/src-app/translations/en/components.json]: done

File: runbotics/runbotics-common/src/model/api/guest.model.ts (hunk 4 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-common/src/model/api/guest.model.ts]: done

File: runbotics/runbotics-scheduler/src/database/guest/guest.entity.ts (hunk 6 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/database/guest/guest.entity.ts]: done

File: runbotics-orchestrator/src/main/java/com/runbotics/service/ProcessService.java (hunk 2 above)
Comment (star-in-the-jar) [reply runbotics-orchestrator/src/main/java/com/runbotics/service/ProcessService.java]: i'll keep that in mind

File: runbotics-orchestrator/src/main/java/com/runbotics/web/rest/ProcessResource.java (hunk 3 above)
Comment (star-in-the-jar) [reply runbotics-orchestrator/src/main/java/com/runbotics/web/rest/ProcessResource.java]: done & done

**Review by preludium — DISMISSED** [review]

**Review by preludium — APPROVED** [review]

**Pull request merged**